### ISSUE TITLE
How to modify <pre> tag

### ISSUE THREAD

**Issue by light-souler** [expressive-code/expressive-code#242]
I use Astro with Expressive-code, this is expressive-code style: 
![image](https://github.com/user-attachments/assets/8a410d32-fb56-4f14-bf5d-acf8f8dbee8d)

```
 styleOverrides: {
      // You can also override styles
      borderWidth: '3px',
      borderRadius: '0rem',
      frames: {
        frameBoxShadowCssValue: 'none',
        editorBackground: '#00000007',
        editorTabBorderRadius: '0',
        
      },
    },
```

1. I have question, how I can add border-style & border-color, like this:
![image](https://github.com/user-attachments/assets/59aeffe6-eb35-45c2-8e59-87319b94d162)

2. I don't want see text: "copied!", can I change behavior, just change icon from [paper] -> [confirmed] ✓

**Comment by hippotastic** [comment]
Hello and welcome! :)

1. You can change the border color of code blocks using the [`borderColor`](https://expressive-code.com/reference/style-overrides/#bordercolor) styleOverrides setting, which you can add to the code you've shown in your question right next to your existing `borderWidth` and `borderRadius` settings. Please note that you can also pick from lots of predesigned themes which come with their own individual border colors, as you can see on the [Themes](https://expressive-code.com/guides/themes/) page.

2. You can currently change the contents of the "Copied!" tooltip in your Astro or EC config file. For example, if you're using `astro-expressive-code`, you can add the following lines to your Astro or EC config file to override the copy button copied text for all locales with a simple checkmark:

   ```js
   import { pluginFramesTexts } from 'astro-expressive-code'
   
   pluginFramesTexts.overrideTexts(undefined, {
     copyButtonCopied: '✔',
   })
   ```

   This will result in the following copy feedback: 
   ![image](https://github.com/user-attachments/assets/064db9d4-7116-4ca7-b15e-456113817598)

   There is currently no built-in option to get rid of the tooltip altogether and replace the icon with a checkmark instead. However, you can still achieve this effect by adding some CSS through a custom plugin, which you can always add to Expressive Code's `plugins` option:
   
   ```js
   plugins: [
     // ... your other plugins ...
     {
       name: 'Custom Copy Button Feedback',
       baseStyles: `
         .copy .feedback {
           display: none;
 
           & + button::after {
             -webkit-mask-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24' fill='black' stroke='none' stroke-width='1.75'%3E%3Cpath d='M9 21.035l-9-8.638 2.791-2.87 6.156 5.874 12.21-12.436 2.843 2.817z'/%3E%3C/svg%3E");
             mask-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24' fill='black' stroke='none' stroke-width='1.75'%3E%3Cpath d='M9 21.035l-9-8.638 2.791-2.87 6.156 5.874 12.21-12.436 2.843 2.817z'/%3E%3C/svg%3E");
           }
         }
       `,
     },
   ],
   ```
   
   This will result in the following copy feedback:
   ![image](https://github.com/user-attachments/assets/354b28e6-8101-4d08-bd22-dc42872d804e)

Note that you can always write your own plugins as shown above to add any custom CSS, even if there is no `styleOverrides` setting for your desired use case.

Hope that helps!

**Comment by light-souler** [comment]
@hippotastic tks you so much, plugin is great :v 
I like simple icon.

I have another question, can I always to show [paper] icon,  
I don't want effect hover <pre> code after show icon [paper] to copy

**Comment by hippotastic** [comment]
Hello again! I'm happy to hear that the plugin worked well for you. :) Here are my responses to your new questions:

### How to always display the copy button

The copy button is already always displayed on devices that don't support hovering (e.g. mobile touch devices). To extend this behavior to all devices, you can add another small plugin:

```js
plugins: [
  // ... your other plugins (including the copy button icon plugin) ...
  {
    name: 'Always Show Copy Button',
    baseStyles: `
      @media (hover: hover) {
        .copy button {
          opacity: 0.75;
        }
      }
    `,
  },
],
```

### How to shorten the "copied" time in the icon-based copy feedback plugin

The time until the icon returns from the copied feedback to the clipboard icon is hardcoded and cannot be changed. However, you can modify the CSS selector in my example icon plugin to make the display time shorter, but that comes at the expense of the "copied" checkmark icon not being shown again when the button is clicked again within 2.5 seconds of the last click. If that is not a problem for you, here is another version of the custom copy button feedback plugin (the only difference is the change from `&` to `&.show` in the icon selector):

```js
plugins: [
  // ... your other plugins ...
  {
    name: 'Custom Copy Button Feedback',
    baseStyles: `
      .copy .feedback {
        display: none;

        &.show + button::after {
          -webkit-mask-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24' fill='black' stroke='none' stroke-width='1.75'%3E%3Cpath d='M9 21.035l-9-8.638 2.791-2.87 6.156 5.874 12.21-12.436 2.843 2.817z'/%3E%3C/svg%3E");
          mask-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24' fill='black' stroke='none' stroke-width='1.75'%3E%3Cpath d='M9 21.035l-9-8.638 2.791-2.87 6.156 5.874 12.21-12.436 2.843 2.817z'/%3E%3C/svg%3E");
        }
      }
    `,
  },
],
```

### How to hide the "Copy to clipboard" tooltip

The "Copy to clipboard" text can be overridden in the same way as I described it for the copied text. In this case, you want to replace the tooltip text with a single space, causing EC to respect it and browsers to ignore it:

```js
// astro.config.mjs
// ... other Astro imports ...

import { pluginFramesTexts } from 'astro-expressive-code'

pluginFramesTexts.overrideTexts(undefined, {
  copyButtonTooltip: ' ',
})

// ...other Astro config code...
```

Hope that helps!

**Comment by light-souler** [comment]
@hippotastic 
Thank you so much, I got great results!

<img width="660" alt="image" src="https://github.com/user-attachments/assets/47b24ce4-b46c-4e8f-95b3-f1a47b61df11">

I will not use &.show because it feels like a trick rather than a proper feature.

You mentioned that the time for the copied feedback is hardcoded. I don’t understand why it takes 3 seconds to return to the default state. This may decrease user satisfaction. When I tested it by clicking the copy button on two different code blocks, the icon didn’t change for 3 seconds (& + button::after), but the copy action still worked.

In the future, I hope we can have a **dynamic response time feature**, like this:
After the text is successfully copied to the clipboard, the [confirmed] icon appears immediately.
If the text is too long, like a file that takes more than 3 seconds to copy, it shows a "loading button" instead for the required duration.

**Comment by hippotastic** [comment]
Happy to hear you got great results!

The only reason why there is a hardcoded delay of 2.5s is to give users enough time to read the default "Copied!" feedback tooltip text. Actual copying never takes longer than a few milliseconds.

The plugin-based solutions I shared with you in this thread just hook into the existing classes and states of the existing copy button functionality to change their visual representation. There is currently no state available that returns quicker to the default that we could hook.

In a future release, I'll provide a built-in option to switch to an icon-based copy feedback as we discussed here, so that EC users can select their preferred method of feedback in the config. This built-in option will then also come with a quicker switch back to the default icon after copying.

Thanks again for your ideas here!
